src/http.cc

```diff
@@ -826,11 +839,111 @@ HttpStateData::handle1xx(HttpReply *reply)
     // for similar reasons without a 1xx response.
 }
 
+/// Matching protocols predicate function for std::find_if algorithm
+/// It handle the case the reference protocol includes a protocol version
+/// but the checking protocol does not (eg when Squid requests upgrade to
+/// 'websockets' and server reports upgrade to 'websocket/1.2').
+class MatchProtocol
+{
+public:
+    explicit MatchProtocol(const char *s, size_t len) :ref(s), refLen(len) {}
+    bool operator() (const SBuf &checking) {
+        const char *p;
+        if (checking.find('/') == SBuf::npos && (p = std::find(ref, ref + refLen, '/')) != ref + refLen)
+            return checking.caseCmp(ref, p - ref) == 0;
+        else
+            return checking.caseCmp(ref, refLen) == 0;
+    }
+
+private:
+    const char *ref;
+    size_t refLen;
+};
+
+bool
+HttpStateData::upgradeProtocolsSupported(const HttpReply *reply) const
+{
+    if (!upgradeProtocols)
+        return false;
+
+    String upgradeProtos = reply->header.getList(Http::HdrType::UPGRADE);
```

Comment:
Please use `const` or 'const auto` if possible.

Reply:
> Would requestedSupportedUpgradeProtocols go?


It is a bad name.


>    the algorithm becomses more clear because it showes when we need to search for the exact match and when just the prefix match.
>    We do not need to maintain and document a special class, which is used only once.

In the other hand it is not case insensitive, it is complex (a lamda function
of three compressed lines is complex) and requires conversion from "const char *"
to SBuf. Also I would use "const bool" instead of "const auto" to avoid
confusing users about what we are trying to do here.

However I restruct MatchProtocol to used bool variables to make more clear what the 
MatchProtocol is doing and avoid avoid recomputing refIsverioned again on each check.

> Ok, then please use const (or const auto for the upgrade parameter.

OK for const